test/micro_ros_utilities/test_memory.hpp

```diff
@@ -0,0 +1,238 @@
+// Copyright (c) 2021 - for information on the respective copyright owner
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef MICRO_ROS_UTILITIES__TEST_MEMORY_HPP_
+#define MICRO_ROS_UTILITIES__TEST_MEMORY_HPP_
+
+#include <gtest/gtest.h>
+
+#include <micro_ros_utilities/type_utilities.h>
+#include <micro_ros_utilities/string_utilities.h>
+
+#include <utility>
+#include <string>
+#include <vector>
+#include <map>
+
+class MicroROSTypeUtilities : public ::testing::Test
+{
+public:
+  MicroROSTypeUtilities()
+  : dyn_msg_(nullptr),
+    static_buffer_(nullptr),
+    allocated_memory_map_()
+  {}
+
+  ~MicroROSTypeUtilities() {}
+
+  void SetUp() override
+  {
+    allocators_ = {
+      allocate,
+      deallocate,
+      reallocate,
+      zero_allocate,
+      static_cast<void *>(this)
+    };
+
+    conf_.allocator = &allocators_;
+
+    ASSERT_EQ(allocated_memory(), 0UL);
+  }
+
+  void TearDown() override
+  {
+    if (static_buffer_ != nullptr) {
+      free(static_buffer_);
+    }
+
+    // Ensure not memory leak in memory handling
+    ASSERT_EQ(allocated_memory(), 0UL);
+
+    // Ensure that realloc and calloc are not used;
+    ASSERT_EQ(reallocate_count, 0UL);
+    ASSERT_EQ(zero_allocate_count, 0UL);
+  }
+
+  // Handle messages
+  template<typename T>
+  bool create_memory_dynamic(T * msg, rcutils_allocator_t * override_allocators = nullptr)
+  {
+    size_t initial_allocated_memory = allocated_memory();
+
+    if (override_allocators != nullptr) {
+      conf_.allocator = override_allocators;
+    }
+
+    allow_dynamic_operations_ = true;
+    EXPECT_NE(msg, nullptr);
+    EXPECT_EQ(dyn_msg_, nullptr);
+
+    dyn_msg_ = msg;
+
+    size_t size = micro_ros_utilities_get_dynamic_size(
+      typesupport_,
+      conf_
+    );
+
+    bool ret = micro_ros_utilities_create_message_memory(
+      typesupport_,
+      msg,
+      conf_);
+
+    if (override_allocators == nullptr) {
+      EXPECT_EQ(allocated_memory() - initial_allocated_memory, size);
+    }
+
+    return ret;
+  }
+
+  void destroy_memory_dynamic()
+  {
+    if (dyn_msg_ != nullptr) {
+      allow_dynamic_operations_ = true;
+      ASSERT_TRUE(
+        micro_ros_utilities_destroy_message_memory(
+          typesupport_,
+          dyn_msg_,
+          conf_)
+      );
+    }
+  }
+
+  template<typename T>
+  bool create_memory_static(T * msg)
+  {
+    allow_dynamic_operations_ = rules_.size() > 0;
```

Comment:
Why this?

Reply:
Because when rules are used, the library always uses dynamic memory to handle the rule string comparison.

Reply:
Yes, but if `rules` is empty and `allow_dynamic_operations_ = true` this value will be set to false.
I was thinking something like this:
```suggestion
    if (!allow_dynamic_operations_ && rules_.size() > 0)
    {
      allow_dynamic_operations_ = true;
    }
```